servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -161,26 +150,29 @@ public void onNext(final ServiceDiscovererEvent<ResolvedAddress> event) {
                         event);
                 final List<Host<ResolvedAddress, C>> activeAddresses =
                         activeHostsUpdater.updateAndGet(RoundRobinLoadBalancer.this, currentAddresses -> {
-                            final List<Host<ResolvedAddress, C>> refreshedAddresses = new ArrayList<>(currentAddresses);
-                            final MutableAddressHost<ResolvedAddress, C> searchHost = new MutableAddressHost<>();
-
-                            searchHost.mutableAddress = event.address();
-                            // Binary search because any insertion is performed at the index returned by the search,
-                            // which is consistent with the ordering defined by the comparator
-                            final int i = binarySearch(refreshedAddresses, searchHost, activeAddressComparator);
-
-                            if (event.isAvailable()) {
-                                if (i < 0) {
-                                    refreshedAddresses.add(-i - 1, new Host<>(event.address()));
-                                }
-                            } else if (i >= 0) {
-                                Host<ResolvedAddress, C> removed = refreshedAddresses.remove(i);
-                                if (removed != null) {
-                                    removed.markInactive();
+                            final List<Host<ResolvedAddress, C>> refreshedAddresses =
+                                    new ArrayList<Host<ResolvedAddress, C>>(currentAddresses);
```

Comment:
It is also expected that the ServiceDiscoverer will filter duplicate events, and there is no concurrent modification outside of `Subscriber#onNext`(other than emptying the list). I think we can assume we won't find any duplicates on addition:

This will also avoid ArrayList copy/move of array elements after it has already been copied in the constructor.
```java
                            final ResolvedAddress addr = requireNonNull(event.address());
                            final List<Host<ResolvedAddress, C>> refreshedAddresses;
                            if (event.isAvailable()) {
                                refreshedAddresses = new ArrayList<>(currentAddresses.size() + 1);
                                refreshedAddresses.addAll(currentAddresses);
                                refreshedAddresses.add(new Host<>(addr));
                            } else if (currentAddresses.isEmpty()) {
                                refreshedAddresses = currentAddresses;
                            } else {
                                refreshedAddresses = new ArrayList<>(currentAddresses.size() - 1);
                                for (Host<ResolvedAddress, C> host : (List<Host<ResolvedAddress, C>>) currentAddresses) {
                                    if (host.address.equals(addr)) {
                                        host.markInactive();
                                    } else {
                                        refreshedAddresses.add(host);
                                    }
                                }
                            }
                            return refreshedAddresses;
```

Reply:
I was thinking about double copying when we add an element. Unfortunately, `refreshedAddresses.addAll(currentAddresses)` does 2 copies internally. So, it will be the same as using `new ArrayList(anotherList)` + `add`.

> It is expected that the ServiceDiscoverer will filter duplicate events

DNS SD currently does not filter duplicates. It takes a List from netty's `resolver.resolveAll` and passes it as-is. I didn't check netty's code but assume it just parses each DnsRecord from the response. If DNS response contains duplicated addresses, we will see them here.

Reply:
> I think we can assume we won't find any duplicates on addition

Actually I think we should allow duplicates which enables folks to effectively add weights to certain hosts. I would prefer not to do these two changes (unsorted and allow dups) in the same PR though. So, ya feel free to send a followup

Reply:
https://github.com/apple/servicetalk/pull/1052